ENH: differential_evolution init with population array fixes #6879

andyfaff opened the pull request:
#6879 asks to be able to initialise the population with a given solution. This PR permits the initialisation of the population with a user supplied array.

andyfaff commented:
@ZacDiggum

ilayn commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -400,7 +408,9 @@ def __init__(self, func, bounds, args=(),
 
         self._nfev = 0
-        if init == 'latinhypercube':
+        if isinstance(init, np.ndarray):
```
I think you should test for strings and assume array-like otherwise (such as list of lists would be a valid candidate if passed through `np.array(...)` ). It is annoying when the argument is rejected though it is a valid array but not specifically a Numpy array.

ZacDiggum commented:
Woohoo! Thanks!

andyfaff commented:
Ping.

ZacDiggum commented:
Hi @andyfaff, I'm still here.
I'm very happy with the changes you made regarding the initial population.
For my own purposes I still use my version of `differential_evolution` which also routes the function value `xk` to the callback as discussed in #6878. Back then this led to #6923 and then we sadly got stuck there.
Maybe we could get #6878 going again?

dlax commented:
IHMO, function parameters that can be of different types (or that do not follow a common interface like, say, "a sequence") is a sign of bad API design. Arguably there's already `seed` so we're just adding more here...
I didn't take a close look to these changes though and probably will not be able to do so soon.

ZacDiggum commented:
Sorry, I meant the best-so-far value of the fitness function that could be used as termination criterion.

ZacDiggum commented:
Hi @andyfaff,
how is it going with this pr and also #6878? Since the discussion in #6923 ended without conclusion, how would you like to proceed? The ideas for enhancement are still useful, right?

andyfaff commented:
@ZacDiggum I think the changes in the other two PR's are mostly orthogonal to this PR and probably need a little more discussion. I think this PR is in a state where it could be merged, but it needs to be reviewed first.
@dlax, @rgommers?

rgommers commented on doc/release/1.0.0-notes.rst:
```diff
@@ -52,6 +52,9 @@ The method `trust-exact` was added to the function `scipy.optimize.minimize`.
 This new trust-region method solves the subproblem with higher accuracy at the cost
 of more Hessian factorizations (compared to dogleg) and is able to deal with indefinite
 Hessians. It seems very competitive against the other Newton methods implemented in scipy.
```
need a blank line here

rgommers commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -106,12 +106,18 @@ def differential_evolution(func, bounds, args=(), strategy='best1bin',
             - 'random'
+            - array specifying subset of the initial population. The array can
+              be of shape ``(len(x),)`` or ``(M, len(x))``, where len(x) is the
+              number of parameters. A total of ``min(M, popsize * len(x))``
```
This ``min(M, ...`` is better left out. I thought it was something complicated for a minute, but it just says that if you select ``M`` to be larger than the total size, it will use the total size. No need to explain that in a docstring.

rgommers replied:
It's also not clear to me why you have two options here. What does ``len(x)`` achieve?

EDIT: this is just a special case right, with ``M=1``? And such a small init size is not recommended? If I understood that correctly, then just removing the 1-D ``len(x)`` option would be good.

andyfaff replied:
Both comments are correct. `(len(x),)` is just a special case for M=1, there are no algorithmic problems with M=1.

andyfaff replied:
How about:

            - array specifying subset of the initial population. The array can
              be of shape ``(len(x),)`` or ``(M, len(x))``, where len(x) is the
              number of parameters. `init` is clipped to
              `bounds` before use.

rgommers commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -106,12 +106,18 @@ def differential_evolution(func, bounds, args=(), strategy='best1bin',
         can improve the minimization slightly.
-    init : string, optional
-        Specify how the population initialization is performed. Should be
+    init : string or array-like, optional
```
nitpick: type identifier should be ``str`` not ``string``

andyfaff commented:
There are several reasons why one might want to supply an initial population:

- you create a tight bunch of initial guesses because you 'know' where the solution is, and you don't want to wait too long for convergence
- you want to exclude various parts of parameter space to start with (e.g. you don't necessarily want latin hypercube).
- you want to try your own initialisation methods.

I also have something more long term in mind. I originally wrote differential evolution code in C and for Igor Pro based on https://doi.dx.org/10.1098/rsta.1999.0469. This paper required lower and upper bounds for each parameter - you need them to initialise a random population. Once you've initialised the population you don't technically require bounds (unless you want to stay within a limit). 
A future development that's possible is dropping the requirement to have lower and upper bounds for each parameter, you could just supply an initial guess / initial population. This would bring the interface to differential_evolution in to line with the other optimisers. However, I'm not sure how to manage the change of the differential_evolution interface. The changes to the internals are relatively straightforward. This future possibility is why I'm reluctant to make the DifferentialEvolutionSolver publicly available at the moment.

andyfaff commented:
Addressed comments.

andyfaff commented:
One test failure is a timeout and the other is unrelated.

ZacDiggum commented:
@andyfaff 
It's funny because I'm using `differential_evolution()` for fitting diffraction data as well by minimizing an error function. In my case it's Kogelnik's coupled wave theory in volume holography. That's why I introduced the 'first guess' in the init.
Making bounds optional is very interesting because one could make a curve / model fitting function like `curve_fit()` that uses` differential_evolution()`.
Ideally, the user could choose to use bounds, first guess or both or none.

andyfaff commented:
Do you know of a way of initializing the population without supplying one,
or having bounds?
(Check out refnx or fitting as well)

On 26 Jul 2017 7:08 pm, "ZacDiggum" <notifications@github.com> wrote:

@andyfaff <https://github.com/andyfaff>
It's funny because I'm using differential_evolution() for fitting
diffraction data as well by minimizing an error function. In my case it's
Kogelnik's coupled wave theory in volume holography. That's why I
introduced the 'first guess' in the init.
Making bounds optional is very interesting because one could make a curve /
model fitting function like curve_fit() that usesdifferential_evolution().
Ideally, the user could choose to use bounds, first guess or both or none.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/scipy/scipy/pull/7115#issuecomment-317995512>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAq51v1n2S_37W0IG8ITsfJBrizp_qYZks5sRwH7gaJpZM4MPaZ9>
.

ZacDiggum commented:
My first naive thought would be to use np.finfo(float).min and .max as pseudo bounds or something like that. But one would have to make sure that these values don't crash the fitness function. What happens with the other minimizers when they approach the domain boundaries of the function to be minimized?

andyfaff commented:
@rgommers, I believe this is ready to go (once tests pass). I think all your comments are addressed.

rgommers commented on doc/release/1.0.0-notes.rst:
```diff
@@ -88,6 +88,10 @@ This new trust-region method solves the subproblem with higher accuracy at the c
 
+The ``init`` keyword for the function `scipy.optimize.differential_evolution`
+can now accept an array. This array allows the user to specify a subset of the
+initial population.
```
duplicate content

rgommers commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -106,17 +106,24 @@ def differential_evolution(func, bounds, args=(), strategy='best1bin',
+        covered. Use of an array to specify a population subset could be used,
+        for example, to create a tight bunch of initial guesses in an location
+        where the solution is known to exist, thereby reducing time for
+        convergence.
```
This "subset" approach is not explained and quite unusual. What happens is that ``latinhypercube`` is still used, but an arbitrary part of the initial population that that generates is then *overwritten* with the provided array. Why not just take the given array as the full initial population.

Note that this also doesn't accomplish the goal stated in the PR of letting users try their own initialization method, it'll be a weird mix of two methods.

andyfaff replied:
I was making the changes for how I would use it. Here are the situations I originally envisaged:

1) You want to initialise the population with e.g. latin hypersquares, but you also have an idea for what the best fit should be (c.f. the initial guess). So to seed the minimization you supply an array of shape `(len(x),)`.
2) You want to seed the population with various initial guesses (e.g. reflecting a covariance matrix)
3) You want to specify the entire population.

In order to achieve 1) you have to have an array to copy into, so that's how I came up with the current design.

Perhaps the following could be used instead:

1) if the initial guess has shape `(len(x),)`, i.e. 1-D, then initialise with latin hypersquares, but only set the initial population member.
2) if the initial guess is 2D, then it's assumed to specify the entire population.

andyfaff commented:
Rebased and squashed.
This PR allows you to supply either a `str` -or- an array to the `init` kwd of the `differential_evolution` function. From the docstring:

```
Specify which type of population initialization is performed. Should be
        one of:

            - 'latinhypercube'
            - 'random'
            - array specifying the initial population. The array should have
              shape ``(M, len(x))``, where len(x) is the number of parameters.
              `init` is clipped to `bounds` before use.

        The default is 'latinhypercube'. Latin Hypercube sampling tries to
        maximize coverage of the available parameter space. 'random'
        initializes the population randomly - this has the drawback that
        clustering can occur, preventing the whole of parameter space being
        covered. Use of an array to specify a population subset could be used,
        for example, to create a tight bunch of initial guesses in an location
        where the solution is known to exist, thereby reducing time for
        convergence.
```

The tests pass and I believe most of the concerns have been addressed. However, @dlax did point out that it was messy to supply either a string or an array to the `init` keyword:

> IHMO, function parameters that can be of different types (or that do not follow a common interface like, say, "a sequence") is a sign of bad API design.

An alternative is to supply the population initialisation array via the `popsize` keyword instead. Thus, if `popsize` is an integer or `np.asarray(popsize).size == 1`, then it's use will be the current behaviour. If it's an `(M, len(x))` array, then it would be the initialisation array.
It shouldn't take too long for me to do those changes.

Opinions please (@bjude). If the current approach is considered best, then this should be good to merge. If the alternative approach is considered best I could refactor quite simply.

bjude commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -400,19 +425,23 @@ def __init__(self, func, bounds, args=(),
-        self.num_population_members = popsize * self.parameter_count
+        # the minimum is 5 because 'best2bin' requires a population that's at
+        # least 5 long
+        self.num_population_members = max(5, popsize * self.parameter_count)
```
A minimum of 5 is enforced here but can be overridden if array initialisation is used with fewer than 5 members.
Suggest adding a check in ``init_population_array()`` that raises an exception if less than 5 members are passed.

bjude commented:
Other than the array initialisation check, im happy.
Now might be a good time for a PR of a  general utility function for latin hypercube sampling so that the user can have arbitrary population size with hypercube sampling (as mentioned by @andyfaff in #8093)

andyfaff commented:
Just need to decide on whether you supply the initial population to the `init` kwd or the `popsize` kwd.

andyfaff commented:
Test failure on travisCI/OSX is unrelated.

bjude commented:
I think ``init`` makes more sense, since we're directly initialising the population. The fact that it affects the population size is secondary IMO.

gusgordon commented:
This is super userful, thanks!

andyfaff commented:
@rgommers, I went with your suggestion that the provided population is used for the actual size (rather than providing a subset).
The only decision to be made is whether the population should be given via the `init` keyword (which currently accepts strings), or via the `popsize` keyword (which currently accepts a single integer and is a multiplier factor). The PR is currently taking the former option, but @dlax commented that mixing string options with an array is poor api choice.
There is a third way, which is to add another keyword option.

Could I have feedback on what you think is the best option? If it's the first, then this PR should be ok to merge (I don't like to merge my own PRs). If it's the 2nd/3rd I can quickly refactor.

rgommers commented:
> The only decision to be made is whether the population should be given via the `init` keyword (which currently accepts strings), or via the `popsize` keyword (which currently accepts a single integer and is a multiplier factor).

`init` certainly sounds preferable to `popsize`. I'd also prefer it to yet another keyword. I agree with @dlax in principle, but in practice I think using strings + array in `init` is only slightly ugly and overall the best solution.

andyfaff commented:
In that case I think it's good to go.

On 3 December 2017 at 19:53, Ralf Gommers <notifications@github.com> wrote:

> The only decision to be made is whether the population should be given via
> the init keyword (which currently accepts strings), or via the popsize
> keyword (which currently accepts a single integer and is a multiplier
> factor).
>
> init certainly sounds preferable to popsize. I'd also prefer it to yet
> another keyword. I agree with @dlax <https://github.com/dlax> in
> principle, but in practice I think using strings + array in init is only
> slightly ugly and overall the best solution.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/7115#issuecomment-348749921>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAq51jmhoV1LX0634_xGojiC9iS2ml9Xks5s8mGjgaJpZM4MPaZ9>
> .
>



-- 
_____________________________________
Dr. Andrew Nelson


_____________________________________

rgommers commented:
Let's give it a day or two to see if no one objects, then hit the green button.

dlax commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -469,6 +498,40 @@ def init_population_random(self):
+
+        if (np.size(popn, 0) < 5 or
+                    popn.shape[1] != self.parameter_count or
+                    len(popn.shape) != 2):
```
Too much indentation here. Should be 8 space right of `if`.

andyfaff replied:
Done

dlax commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -469,6 +498,40 @@ def init_population_random(self):
+                                    np.inf)
+
+        # reset number of function evaluations counter
+        self._nfev = 0
```
There's some copy/paste of other `init_population_*` methods here. Not a blocker for this change, but could be worth refactoring a bit.

dlax commented on scipy/optimize/_differentialevolution.py:
```diff
@@ -469,6 +498,40 @@ def init_population_random(self):
+        if (np.size(popn, 0) < 5 or
+                    popn.shape[1] != self.parameter_count or
+                    len(popn.shape) != 2):
+            raise ValueError(self.__init_error_msg)
```
Not sure this is a good idea to use the same message here and in `__init__` since these are arguably different errors.

andyfaff replied:
Done.

dlax commented:
> @andyfaff pushed 40 commits.

Ugh. What's happening here?

andyfaff commented:
Yeah I know. When I checked out the branch on my computer git said my local version and remote branches had diverged, so I pulled the remote version onto mine. I made the changes, and pushed back to the remote and this happened.
I'm going to rebase on master and see if that fixes things. If not then I might use the repo I have on another computer.

andyfaff commented:
Should be fixed
